mx.py

```diff
@@ -2376,7 +2376,7 @@ def _load_libraries(self, libsMap):
             deps = Suite._pop_list(attrs, 'dependencies', context)
             path = attrs.pop('path', None)
             urls = Suite._pop_list(attrs, 'urls', context)
-            sha1 = attrs.pop('sha1', None)
+            sha1 = mx_urlrewrites.rewritesha1(urls, attrs.pop('sha1', None))
```

Comment:
> That way we have (1) precise control over which values (rewritten or not) are passed to _get_path_in_cache and (2) all the rewriting is done in one place.

alternatively, we could move all the rewriting into `_load_libraries` which would have the same affect.

Reply:
> we could move all the rewriting into _load_libraries which would have the same affect

That sounds like the best approach to me as it prevents the un-rewritten urls from leaking out anywhere. Can you please try that change @ceresek .

Reply:
Thank you for the tips, did the update in `Suite::_load_libraries`. More specifically:
- removed rewrites inside `Library` (`__init__` and `get_urls`)
- removed rewrites inside `ResourceLibrary` (`__init__` and `get_urls`)
- implemented replacement rewrites inside `Suite::_load_libraries`

There are few remaining bits to maybe consider:
- original rewriting in resource classes only took place after `substVars` which is now not the case. If we need that (my use case does not) then the rewriting probably has to happen later, inside the classes again. (Also, my earlier patch moved `substVars` so that it happened in the constructor rather than in `get_urls` methods, not sure if this can matter.)
- `Library::get_source_path` rewrites source URL which I did not touch. I can add rewriting of source URL lists into `_load_libraries` too, and move it out of the `get_source_path` method, but by use case does not require that.
- `Repository::get_url` does optional rewrite which I did not touch. Maybe it is now reduntant.
- `parse_specification` and `get_source_urls` in `SuiteImport` do rewrites which I did not touch.

Does this work for you ? (In particular, it should be tested on other scenarios that use URL rewrites, and the tests should make sure that rewriting really happens, I'm not sure if there are code paths that construct resource classes without going through `_load_libraries`.)

Reply:
> * original rewriting in resource classes only took place after `substVars` which is now not the case. If we need that (my use case does not) then the rewriting probably has to happen later, inside the classes again. (Also, my earlier patch moved `substVars` so that it happened in the constructor rather than in `get_urls` methods, not sure if this can matter.)

Nice catch! That is a problem. [This](https://github.com/oracle/graal/blob/e9c1e39393be26d2920a5394b94d41f7567cceca/sdk/mx.sdk/suite.py#L147-L184) for example depends on substitutions before rewriting. :disappointed: 

> * `Library::get_source_path` rewrites source URL which I did not touch. I can add rewriting of source URL lists into `_load_libraries` too, and move it out of the `get_source_path` method, but my use case does not require that so I do not have a way to test it.

I'd move it to `_load_libraries`, but then the `substVars` issue applies as well.

> * `Repository::get_url` does optional rewrite which I did not touch. Maybe it is now redundant.
> * `parse_specification` and `get_source_urls` in `SuiteImport` do rewrites which I did not touch.

IMHO it is fine to keep those, at least for now.

The `substVars` is a blocker for this approach IMHO. The existing code is a bit of mess, sorry about that.

My current understanding is that: `_get_path_in_cache` requires `urlrewrites` which requires `substVars` which probably needs the constructor of the super class to be called to initialize the member variables. That would mean we should move the `_get_path_in_cache` code into the constructors. In addition, I think we should do the `substVars` and `urlrewrites` of `source_urls` also there. I think it does not make sense to store the un-substituted, un-rewritten source urls next to substituted and rewritten download urls.

That is quite a bit of refactoring. :neutral_face: 

@ceresek if you are actively blocked by the missing feature, I'd also be fine with pushing your original suggestion and clean that up later.
@dougxc thoughts?

Reply:
Your suggestions sound good to me @zapster and I'd really prefer to take the opportunity to clean this up now rather than leave another "broken window" lying around to bite us later. @ceresek would you be able to try make the suggested changes?

Reply:
@zapster @dougxc Here is a version that fixes the original issue without much refactoring. In the end, I opted to keep `_load_libraries` and `_get_path_in_cache` without change, the rewriting happens entirely inside the library classes.

This change felt safer than the refactoring you mention - I simply was not sure I would get all the corner cases right.

Can you please take a look ?

Reply:
Modulo the comment I just left about leaking values outside the `Library` constructor, I think your changes are fine as is. What do you think @zapster ?

Reply:
Agreed